design/020-full_dependency_graph_with_build_requires.md

```diff
@@ -0,0 +1,57 @@
+- It makes the Conan codebase much more complicated, and the treatment of dependencies more heterogeneous, which makes thorough and flexible management of dependencies more difficult. The extra maintenance slows down development.
+
+
+## Proposal
+
+When the dependency graph is computed, the whole dependency graph will be computed. The algorithm could be outlined as:
+
+1. Starting from the initial conanfile, compute the dependencies and transitive dependencies of ``requires`` and ``build_requires``. We could have at this stage a ``conanfile.txt``->``poco/1.9.4``->``zlib/1.2.11``, and also ``poco/1.9.4``-(build-require)->``cmake/3.14``
+2. Compute which packages need to be built, based on the current existing binaries and the the ``--build`` policies
+3. Install the binary packages, downloading pre-compiled binaries, finding them already in the cache or building them from sources
+
+
+## Detailed Design
+
+This approach would solve the above issues. It will have some other potential issues:
+
+- It might be slightly less efficient, as it might need to resolve some extra recipes as ``cmake/3.14`` that wouldn’t be strictly necessary otherwise. But only the recipes will be necessary, downloading and unzipping the potentially heavy binary packages can still be avoided, so the impact on the overall time might be negligible. In practice, it could suppose and average penalty of around 1-2% of the installation of a graph with existing binaries (much less if something is built from sources)
+- It requires the ``build_require`` recipes to be accessible and available in different stages, for example, if recipes are copied from one server to another one, it is necessary to copy too those ``build_requires``, if later we plan to do a ``conan install``. Note that if the ``build_require`` is injected via the profile, it would be enough to not pass it in the profile, and it will not be necessary (but in this case the binaries ``package_id`` should be independent of the ``build_requires``). Also note that in many cases this is not necessarily a disadvantage, but desired behavior, as it allows full build reproducibility from the new server too.
+- It can resolve to invalid binaries in some configurations, but this can be a non-fatal condition. For example, when cross-building. Let's say that we have a cross-compiler "mygcc-arm/1.2" that works in Windows and can cross-compile for Linux arm architecture. If this is a ``build_require``, in Windows we could use a ``conan create . -pr:h=LinuxArm -pr:b=Windowsx86`` with the build and host profiles. Once the binaries are built, we could go to a Linux box and do ``conan install … -pr:h=LinuxArm``, but the binary for the cross-compiler “mygcc-arm/1.2” will not exist for that platform, again this is not necessarily an error.  The binary can simply be marked as INVALID by the ``validate()`` method, and unless we tried to build, it would still be possible to install and use the desired binaries.
```

Comment:
> but the binary for the cross-compiler “mygcc-arm/1.2” will not exist for that platform,

It's not that it doesn't exist. It's that the build requirement is completely missing. So I assume hash will be different. Or do I miss something?

Reply:
No, the ``build_requirement`` **declaration** is technically there. For most situations it is more than enough to know the version of it, in that example, lets say that you know that you want to rebuild binaries if your compiler bumps the major, to something like ``mygcc-arm/2.0``. Then you declare a ``major_mode`` on that build-requirement. 

If you are installing in Linux, you will not have a binary that can run in Linux, but you will still know the version, even down to the recipe revision of ``mygcc-arm/1.2`` and you will know that you don't need new binaries, and don't need to fetch the whole ``mygcc-arm/1.2`` package binaries because you don't need to build. But you know that the version is ``1.2``, and everything fits, with just the declaration of that package without needing its full binaries. 

Please let me know if this clarifies it.

Reply:
Not really.
When I build package with 2 profiles I have a bunch of build requirements coming from it.
When I consume a package I don't specify any build profile therefore I have no build requirements => different graph => different package id.

Reply:
That is up to you. If you want that behavior, then you can have it, as by default build_requirements will not be automatically added to the ``package_id``. Basically do nothing, and you will not have a different ``package_id`` when your build_requires are added by a profile or when they are not. If at some point you want to control it and decide that changing some of your build_requires and upgrading it to a newer version should automatically guarantee that packages using that build requires must be rebuilt, then you can add it to the ``package_id`` policy. From that time, yes, you should declare your build_requires in your profiles even if you don't intend to build, because that declaration is necessary to define which binaries to install.